pkg/protocol/engine/consensus/epochgadget/epochorchestrator/performance/performance.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/iotaledger/hive.go/kvstore"
+	"github.com/iotaledger/hive.go/lo"
+	"github.com/iotaledger/iota-core/pkg/core/account"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/blocks"
+	"github.com/iotaledger/iota-core/pkg/storage/prunable"
+	iotago "github.com/iotaledger/iota.go/v4"
+)
+
+type Tracker struct {
+	rewardBaseStore kvstore.KVStore
+	poolStatsStore  kvstore.KVStore
+	committeeStore  kvstore.KVStore
+
+	performanceFactorsFunc func(slot iotago.SlotIndex) *prunable.PerformanceFactors
+
+	timeProvider  *iotago.TimeProvider
+	decayProvider *iotago.ManaDecayProvider
+
+	performanceFactorsMutex sync.RWMutex
+	mutex                   sync.RWMutex
+}
+
+func NewTracker(
+	rewardsBaseStore kvstore.KVStore,
+	poolStatsStore kvstore.KVStore,
+	committeeStore kvstore.KVStore,
+	performanceFactorsFunc func(slot iotago.SlotIndex) *prunable.PerformanceFactors,
+	timeProvider *iotago.TimeProvider,
+	decayProvider *iotago.ManaDecayProvider,
+) *Tracker {
+	return &Tracker{
+		rewardBaseStore:        rewardsBaseStore,
+		poolStatsStore:         poolStatsStore,
+		committeeStore:         committeeStore,
+		performanceFactorsFunc: performanceFactorsFunc,
+		timeProvider:           timeProvider,
+		decayProvider:          decayProvider,
+	}
+}
+
+func (t *Tracker) RegisterCommittee(epoch iotago.EpochIndex, committee *account.Accounts) error {
+	return t.storeCommitteeForEpoch(epoch, committee)
+}
+
+func (t *Tracker) BlockAccepted(block *blocks.Block) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	t.performanceFactorsMutex.Lock()
+	defer t.performanceFactorsMutex.Unlock()
+
+	// TODO: check if this block is a validator block
+
+	performanceFactors := t.performanceFactorsFunc(block.ID().Index())
+	pf, err := performanceFactors.Load(block.Block().IssuerID)
+	if err != nil {
+		panic(err)
+	}
+
+	err = performanceFactors.Store(block.Block().IssuerID, pf+1)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (t *Tracker) ApplyEpoch(epoch iotago.EpochIndex, committee *account.Accounts) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	epochSlotStart := t.timeProvider.EpochStart(epoch)
+	epochSlotEnd := t.timeProvider.EpochEnd(epoch)
+
+	profitMargin := calculateProfitMargin(committee.TotalValidatorStake(), committee.TotalStake())
+	poolsStats := PoolsStats{
+		TotalStake:          committee.TotalStake(),
+		TotalValidatorStake: committee.TotalValidatorStake(),
+		ProfitMargin:        profitMargin,
+	}
+
+	if err := t.poolStatsStore.Set(epoch.Bytes(), lo.PanicOnErr(poolsStats.Bytes())); err != nil {
+		panic(errors.Wrapf(err, "failed to store pool stats for epoch %d", epoch))
+	}
+
+	committee.ForEach(func(accountID iotago.AccountID, pool *account.Pool) bool {
+		intermediateFactors := make([]uint64, 0)
+		for slot := epochSlotStart; slot <= epochSlotEnd; slot++ {
+			performanceFactorStorage := t.performanceFactorsFunc(slot)
+			if performanceFactorStorage == nil {
+				intermediateFactors = append(intermediateFactors, 0)
```

Comment:
seems like there is a `continue` missing here?
otherwise the next line would panic IMO